rs/replicated_state/src/canister_state/system_state.rs

```diff
@@ -268,6 +271,221 @@ impl CanisterHistory {
     }
 }
 
+/// `TaskQueue` represents the implementation of queue structure for tasks satisfying  conditions:
+///
+/// 1. If there is a `Paused` or `Aborted` task it will be returned first.
+/// 2. If an `OnLowWasmMemoryHook` is ready to be executed, it will be returned next.
+/// 3. All other tasks will be returned based on the order in which they are added to the queue.
+#[derive(Clone, Eq, PartialEq, Debug, Default)]
+pub struct TaskQueue {
+    /// Queue of tasks.
+    queue: VecDeque<ExecutionTask>,
+
+    /// Status of low_on_wasm_memory hook execution.
+    on_low_wasm_memory_hook_status: OnLowWasmMemoryHookStatus,
+}
+
+impl TaskQueue {
+    pub fn from_checkpoint(
+        queue: VecDeque<ExecutionTask>,
+        on_low_wasm_memory_hook_status: OnLowWasmMemoryHookStatus,
+    ) -> Self {
+        TaskQueue {
+            queue,
+            on_low_wasm_memory_hook_status,
+        }
+    }
+
+    pub fn front(&self) -> Option<&ExecutionTask> {
+        self.queue.front()
+    }
+
+    pub fn pop_front(&mut self) -> Option<ExecutionTask> {
+        self.queue.pop_front()
+    }
+
+    #[cfg(test)]
+    pub fn pop_back(&mut self) -> Option<ExecutionTask> {
+        self.queue.pop_back()
+    }
+
+    pub fn push_front(&mut self, task: ExecutionTask) {
+        match task {
+            ExecutionTask::OnLowWasmMemory(hook_condition_check_result) => {
+                if let Some(is_hook_condition_satisfied) = hook_condition_check_result {
+                    self.on_low_wasm_memory_hook_status
+                        .update(is_hook_condition_satisfied);
+                }
+            }
+            ExecutionTask::Heartbeat
+            | ExecutionTask::GlobalTimer
```

Comment:
Should we allow `push_front()` for anything other than `Paused` or `Aborted` tasks? What's the point of wrapping all the state inside `TaskQueue` if we allow callers to trivially break the very invariants listed in the documentation of the struct?

I suppose that `OnLowWasmMemory` is fine, since we're not actually enqueuing it. But `Heartbeat` and `GlobalTimer` should definitely go to the back of the queue.

It's likely best if we just name this method `enqueue()` and the implementation decides whether to enqueue at the front; at the back; or on the side, in `self.on_low_wasm_memory_hook_status` (and later in the field holding the only paused/aborted task).

Reply:
The name is changed to `enqueue`.